### PR TITLE
`struct Rav1dFrameContext`: Separate into data and context

### PR DESCRIPTION
**Pull request by rinon**
Separates the frame data into intialized and then rarely updated data (`Rav1dFrameData`) and a few of the fields required to pick tasks before accessing the rest of the frame data to execute the task.

### CODE REVIEW

File: src/thread_task.rs [hunk 1]
```diff
@@ -323,24 +331,29 @@ unsafe fn create_filter_sbrow(
...
     let num_tasks = (f.sbh * (1 + uses_2pass)) as usize;
     tasks.grow_tasks(num_tasks);
     let task_idx = Rav1dTaskIndex::Task((f.sbh * (pass & 1)) as usize);
     if pass & 1 != 0 {
-        f.frame_thread_progress.entropy = AtomicI32::new(0);
+        fc.frame_thread_progress
+            .try_read()
+            .unwrap()
+            .entropy
+            .store(0, Ordering::Relaxed);
```
Comment (randomPoison) [inline src/thread_task.rs:344]: It seems odd to have an atomic inside of an `RwLock`ed struct. Both give us interior mutability, and doing it this way means we're paying twice to mutate that field. Would it make sense to hoist the atomic field out of the `RwLock`?

**Review by randomPoison — COMMENTED** [review]
I gave this a first look over and had a few comments. I can give this another pass when it's ready for final review.

File: src/thread_task.rs (hunk 1 above)
Comment (rinon) [reply src/thread_task.rs:344]: I could move the RwLock down to just the Vecs, we only need it to be able to resize the Vec.

Comment (kkysen) [reply src/thread_task.rs:344]: The relaxed atomic load/store is pretty much free, though, so I don't know if it's the biggest deal.

File: src/thread_task.rs [hunk 2]
```diff
@@ -923,6 +949,7 @@ pub unsafe fn rav1d_worker_task(c: &Rav1dContext, task_thread: Arc<Rav1dTaskCont
         tasks[t_idx].next = None;
         let tile_tasks = tasks.tile_tasks;
         let t = &mut tasks[t_idx];
+        // dbg!(&*t);
```
Comment (randomPoison) [inline src/thread_task.rs:952]: Unused debug comment that can be removed

File: src/thread_task.rs [hunk 3]
```diff
@@ -1031,23 +1064,25 @@ pub unsafe fn rav1d_worker_task(c: &Rav1dContext, task_thread: Arc<Rav1dTaskCont
                                 p_0 += 1;
                             }
                             assert!(task_thread_lock.is_none(), "thread lock should not be held");
+                            // drop(fc_locked);
```
Comment (randomPoison) [inline src/thread_task.rs:1067]: Unused comment can be removed

File: src/thread_task.rs [hunk 4]
```diff
@@ -1186,11 +1225,12 @@ pub unsafe fn rav1d_worker_task(c: &Rav1dContext, task_thread: Arc<Rav1dTaskCont
                                     t.type_0 = TaskType::Cdef;
                                     t.deblock_progress = 0 as c_int;
                                     t.recon_progress = t.deblock_progress;
-                                    add_pending(f, t_idx);
+                                    add_pending(fc, t_idx);
                                     assert!(
                                         task_thread_lock.is_none(),
                                         "thread lock should not be held"
                                     );
+                                    // drop(fc_locked);
```
Comment (randomPoison) [inline src/thread_task.rs:1233]: Unused comment

File: src/thread_task.rs [hunk 5]
```diff
@@ -1319,23 +1364,25 @@ pub unsafe fn rav1d_worker_task(c: &Rav1dContext, task_thread: Arc<Rav1dTaskCont
                 }
             }
             if sby + 1 == sbh {
-                f.task_thread.done[0].store(1, Ordering::SeqCst);
+                fc.task_thread.done[0].store(1, Ordering::SeqCst);
             }
             assert!(task_thread_lock.is_none(), "thread lock should not be held");
+            // drop(fc_locked);
```
Comment (randomPoison) [inline src/thread_task.rs:1370]: unused comment

**Review by randomPoison — APPROVED** [review]
Looks good (modulo any changes necessary to fix the last bit of lock contention), my only comments were minor ones

File: src/decode.rs [hunk 6]
```diff
@@ -4599,52 +4617,52 @@ pub(crate) unsafe fn rav1d_decode_frame_exit(
...
+        res = rav1d_decode_frame_init_cdf(c, fc, &mut f, &fc.in_cdf());
     }
     // wait until all threads have completed
     if res.is_ok() {
         if c.tc.len() > 1 {
-            res = rav1d_task_create_tile_sbrow(c, f, 0, 1);
-            let mut task_thread_lock = (*f.task_thread.ttd).delayed_fg.lock().unwrap();
-            (*f.task_thread.ttd).cond.notify_one();
+            res = rav1d_task_create_tile_sbrow(c, fc, &f, 0, 1);
+            drop(f);
```
[the remark this answers is not in the document]
Comment (rinon) [reply src/decode.rs:4637]: Holding the lock for FrameData while trying to wait on the task thread lock caused contention. We need to drop the FrameData lock to allow other threads to make progress.

[the remark this answers is not in the document]
Comment (rinon) [reply src/decode.rs]: I added a comment to this effect.

**Comment by rinon** [comment]
There is still lock contention at https://github.com/memorysafety/rav1d/blob/e2b43173887f12105be297417c66e33b83009606/src/decode.rs#L4584 that I can't replicate reliably enough to find. I'm going to do some experimentation to see if I can't figure out what's happening here. The alternative is to actually just block there, which might be the best alternative.

**Comment by kkysen** [comment]
> There is still lock contention at
> 
> https://github.com/memorysafety/rav1d/blob/e2b43173887f12105be297417c66e33b83009606/src/decode.rs#L4584
> 
> that I can't replicate reliably enough to find. I'm going to do some experimentation to see if I can't figure out what's happening here. The alternative is to actually just block there, which might be the best alternative.

So this is just for destructors mostly?  I think taking the lock might be okay then.  We should check with some benchmarks and mark it with a comment, but if the perf looks okay I think it's fine.

**Comment by rinon** [comment]
Local benchmark looks ok, I'm going to land this.

**Pull request merged**